flink-runtime/src/main/java/org/apache/flink/runtime/state/track/BackendStateRegistry.java

```diff
@@ -0,0 +1,298 @@
+            if (entry != null) {
+                entry.notActivelyUsed(
+                        backendId,
+                        new TreeSet<>(pendingSavepoints),
+                        new TreeSet<>(pendingCheckpoints));
+                trackNotActivelyUsedEntry(entry);
+            }
+        }
+    }
+
+    private void trackNotActivelyUsedEntry(StateEntry<K> entry) {
+        if (entry.isUsedInSavepoints(backendId)) {
+            LOG.debug(
+                    "State entry not used but {} pending savepoints exist, backend: {}, state: {}",
+                    pendingSavepoints.size(),
+                    backendId,
+                    entry);
+            inUseBySavepoints.put(entry.getKey(), entry);
+        } else if (trackIfUsedByCheckpoints(entry)) {
+            LOG.debug(
+                    "State entry not used but {} pending checkpoints exist, backend: {}",
+                    pendingCheckpoints.size(),
+                    backendId);
+        } else {
+            LOG.debug("State entry not used, backend: {}, state: {}", backendId, entry);
+        }
+    }
+
+    public void checkpointStarting(long checkpointId, boolean isSavepoint) {
+        LOG.debug(
+                "Checkpoint started, backend: {}, checkpoint: {}, isSavepoint: {}",
+                backendId,
+                checkpointId,
+                isSavepoint);
+        Set<Long> ids = isSavepoint ? pendingSavepoints : pendingCheckpoints;
+        checkState(
+                lastStartedCheckpoint < checkpointId,
+                "Out of order checkpoint: %s, backend: %s, previous: %s, isSavepoint: %s",
+                checkpointId,
+                backendId,
+                lastStartedCheckpoint,
+                isSavepoint);
+        ids.add(checkpointId);
+        lastStartedCheckpoint = checkpointId;
+    }
+
+    public void snapshotTaken(
+            long checkpointId, StateObject state, boolean inferUnusedFromPrevious) {
+        Set<K> newStateKeys = keyExtractor.apply(state).keySet();
+        LOG.debug(
+                "Checkpoint performed, backend: {}, checkpoint: {}, inferUnusedFromPrevious: {}, state objects: {}",
+                backendId,
+                checkpointId,
+                inferUnusedFromPrevious,
+                newStateKeys);
+        if (pendingSavepoints.remove(checkpointId)) {
+            snapshotSentForSavepoint(checkpointId, newStateKeys);
+        } else if (pendingCheckpoints.contains(checkpointId)) {
+            // note that we don't remove this checkpoint from pending for now - wait
+            // until subsumed
+            snapshotSentForCheckpoint(checkpointId, inferUnusedFromPrevious, newStateKeys);
+        } else {
+            throw new IllegalStateException(
+                    String.format("Unknown checkpoint %d for backend %s", checkpointId, backendId));
+        }
+    }
+
+    private void snapshotSentForSavepoint(long checkpointId, Set<K> newStateKeys) {
```

Comment:
From my point of view,  the name of `snapshotSentForSavepoint` is a bit hard to understand, `who` `send` the information to `where`?

Reply:
Actually the snapshot is created, not necessarily sent.
How about `savepointSnapshotCreated()`? (and `snapshotTaken` -> `snapshotCreated`)

Reply:
Do you think we could have `global savepoint`, `global checkpoint` and `local snapshot` terminologies to distinguish them?

Reply:
I think both checkpoints and checkpoints are currently global - tasks do not checkpoint independently of each other.
The snapshot can be global or local; but in TM, it's always local (unless everything fits into a single TM :)
So here it will be`local savepoint snapshot` or `local checkpoint snapshot` - and `local` seems redundant.
Or did I get you wrong?

Reply:
The task could have `local snapshot` successful after getting the state handle to return. And the `global checkpoint` might finally success or failed in the end, that's the duty of notification of checkpoint completed or aborted.
From this PR, I can see many place depends on the notifications, and if we could distinguish them, the code could look more clear.